Port to Python3|TelepathyGlib

Saumya-Mishra9129 opened the pull request:
@quozl I have tested the activity it is working properly . The shell.log contains `Normal Successfull completion, pid 1026 activity_id `7d2d4f14a8afe2e95921ba35e602ec40c6bfc7a2`. Kindly review it.

quozl commented on SearchActivity.py:
```diff
@@ -1,3 +1,4 @@
+from builtins import str
```
Copyright should be before code.  Why is this import necessary?

Saumya-Mishra9129 replied:
I will remove this one.

quozl commented on SearchActivity.py:
```diff
@@ -22,21 +25,22 @@
 
 import json
-from json import load as jload
+from json import load as jloadT
```
Why is this import renamed?

Saumya-Mishra9129 replied:
It seems to be happened by mistake. Sorry I'll take care of it next time.

quozl commented on game.py:
```diff
@@ -37,7 +41,7 @@
 
 
-class Game():
+class Game(object):
```
Why is this necessary?

Saumya-Mishra9129 replied:
https://portingguide.readthedocs.io/en/latest/classes.html -have a look at this. It seems it only necessary when we need compatibility with both python 2 and 3.

quozl commented on game.py:
```diff
@@ -67,11 +71,11 @@ def __init__(self, canvas, parent=None, path=None,
             self.grid_width = TEN
-        self._scale = min(self._width / (self.grid_width * DOT_SIZE * 1.2),
-                          self._height / (self.grid_height * DOT_SIZE * 1.2))
+        self._scale = min(old_div(self._width, (self.grid_width * DOT_SIZE * 1.2)),
```
Why use old_div?

quozl reviewed (changes requested):
Thanks.

How did you test collaboration?

Can you work toward making sprites.py consistent with sprites.py in other activities that have been ported?  There's no reason to make this one different, and some good reasons to make it the same.

ayushnawal commented:
Thanks @Saumya-Mishra9129 

You can have a look at `deducto` and review the changes made by @chimosky in sprites.py.

As @quozl pointed out, you have used old_div, in spite of this you could have simply used `//`.\
Although both give the same result `//` is what we used while porting an activity from python 2 to 3.

You can test collaboration by setting up two sugar environments and confirm that you can [collaborate](https://github.com/sugarlabs/help-activity/blob/master/source/collaborating.rst)

Also, there is already a pull request  #20 with the title `Port to Python 3`, I have not reviewed that completely but some changes were similar, I hope there will be no conflict when @quozl merges.

quozl commented:
Thanks for comments.

> As @quozl pointed out, you have used old_div, in spite of this you could have simply used `//`.
> Although both give the same result `//` is what we used while porting an activity from python 2 to 3.

`old_div` is for writing code compatible with both Python 2 and Python 3.  Only the Sugar Toolkit is required to do that.  Activities are mostly either Python 2 or Python 3.

It is best to test to find the type of the arguments given to the division.  Use `/` if one or more of the arguments were floating point when tested on Python 2.  Use `//` if there's a need to avoid a floating point result on Python 3.  Reminds us again why you must be able to test the activity on Python 2 before porting.

> Also, there is already a pull request #20 with the title `Port to Python 3`, I have not reviewed that completely but some changes were similar, I hope there will be no conflict when @quozl merges.

It is nice to have more than one person working on a problem, but I do expect more cooperation and sharing of the patches.  Unless patches are shared between the pull requests, then whichever pull request is ready to merge **will merge first**, then the other pull request will be considered as additional and may need to rebase.  That's why it is very helpful for the two developers who began the two pull requests to review and merge each other's work into their own branches, rather than try to work apart.

Saumya-Mishra9129 commented:
I will take care of cooperation next time. Is there any other activity rather than Sugar Toolkit which needs compatibility with both python 2 and 3? I was using old_div so that division operator work fine with python3 as with python2. I was using futurize as a tool for conversion from python 2 to python 3. I guess I should use 2to3 tool only.

ayushnawal commented:
> Is there any other activity rather than Sugar Toolkit which needs compatibility with both python 2 and 3?

In my opinion, No. Sugar Toolkit needs compatibility with both python 2 and 3 because some activities are running on python3 and some are still on python2( like this activity). So in order to support both the type of activities, the sugar toolkit is made compatible for both 2 and 3.

>I was using futurize as a tool for conversion from python 2 to python 3

I don't know much about futurize and it's advantages over `2to3` tool but the porting guide suggests to use 2to3.

Have a look at [Porting Guide](https://github.com/sugarlabs/sugar-docs/blob/master/src/python-porting-guide.md ) (in case you haven't :grin: )

Thanks

quozl commented:
> I will take care of cooperation next time. Is there any other activity rather than Sugar Toolkit which needs compatibility with both python 2 and 3?

No, not that I know of.

I did consider it when planning the port for all of Sugar, because I'm to maintain Python 2 branches for use on the OLPC XO laptops which are stuck on Python 2 (not your problem).  But the amount of change in activities has not been sufficient to justify the effort.  Instead I backport patches as needed from _master_ to _python2_ branch.

sanjay237 commented:
> `old_div` is for writing code compatible with both Python 2 and Python 3. Only the Sugar Toolkit is required to do that. Activities are mostly either Python 2 or Python 3.

Don't want to be _that_ dude, but the // operator works the same way on Python 2 and Python 3.

quozl commented:
> Don't want to be that dude, but the // operator works the same way
> on Python 2 and Python 3.

Well said, Sanjay, thanks.

(Aside; a trouble with discussing style and porting in pull requests
is that not everyone who has an opinion is present.  Best in future if
challenges to our style are discussed on sugar-devel@).

Saumya-Mishra9129 commented:
@quozl  @sanjay237 thanks for suggestions.

JuiP commented:
> I was using futurize as a tool for conversion from python 2 to python 3. I guess I should use 2to3 tool only.

Almost every output of `2to3` will need modification to provide backward compatibility with Python 2. On the other hand, the `python-future` project provides a script called `futurize` that will produce code that is compatible with both Python 2 and Python 3. I hope I cleared your confusion why you should use `2to3`.

Saumya-Mishra9129 commented:
I have tried to fix Collaboration of activity as I am trying to port the activity to Collabwrapper , the port is still under process but yet I have made some changes. A review is requested here to proceed further as I am still getting some D-Bus connection related errors.

quozl commented:
In general it looks kinda right, but you need to get it working before it is worth doing a detailed review.  If you face an error you cannot resolve, do ask about it.

chimosky commented on SearchActivity.py:
```diff
@@ -37,12 +38,18 @@
 from json import dump as jdump
 from io import StringIO
 
+try:
+    from sugar3.presence.wrapper import CollabWrapper
+except (ImportError, ModuleNotFoundError):
+    from textchannelwrapper import CollabWrapper
```
`CollabWrapper` is maintained independently [here](https://github.com/sugarlabs/collabwrapper) and isn't part of the toolkit, the documentation was to show how to use it.

You can just use `collabwrapper.py` instead of `textchannelwrapper`.

chimosky commented on SearchActivity.py:
```diff
@@ -117,17 +146,57 @@ def _setup_toolbars(self):
         self._game.new_game()
 
+    def _message_cb(self, collab, buddy, msg):
+        ''' Data from a tube has arrived. '''
```
You can also remove or change this doc string as it's misleading because a text channel is used and not a tube.

chimosky commented on SearchActivity.py:
```diff
@@ -117,17 +146,57 @@ def _setup_toolbars(self):
+
+    def send_join(self):
+        _logger.debug('send_join')
+        self.send_event("j", self.nick)
```
You don't need to define a `send_event` that runs a check as `post` only broadcasts a message if activity is shared.

chimosky commented on SearchActivity.py:
```diff
@@ -185,144 +254,61 @@ def _write_scores_to_clipboard(self, button=None):
+                payload=payload
+            ))
 
     def send_new_game(self):
```
You can also remove this function and `post` the needed message whenever this was called.

chimosky commented:
I agree with @quozl above, I took a look and made some comments.

Saumya-Mishra9129 commented:
> I agree with @quozl above, I took a look and made some comments.

Thanks
@chimosky Can you review these pull requests once.
[#46](https://github.com/sugarlabs/physics/pull/46)[](url)
[#79](https://github.com/sugarlabs/turtleart-activity/pull/79)
[#3](https://github.com/sugarlabs/solar-system/pull/3)
[#9](https://github.com/sugarlabs/ruler/pull/9)

Saumya-Mishra9129 committed:
Run 2to3

Saumya-Mishra9129 committed:
Change exec to sugar-activity3

Saumya-Mishra9129 committed:
Port to TelepathyGLib

Saumya-Mishra9129 committed:
Remove Trailing semicolon in the statement

Saumya-Mishra9129 committed:
Update sprites.py with upstream

Source - https://github.com/sugarlabs/turtleart-activity/blob/master/TurtleArt/sprites.py

Saumya-Mishra9129 committed:
Add PangoCairo and Gdk required versions

Saumya-Mishra9129 committed:
Fix Gtk Clipboard TypeError

add len argument in Gtk.Clipboard set_text method
set len=-1

Fixes Error:
TypeError: Gtk.Clipboard.set_text() takes exactly 3 arguments (2 given)

Saumya-Mishra9129 commented:
I am making new commits here as previous commits and port was based on futurize tool , and It makes difficult in reviewing as well as it creates a lot of confusion. I am done with Port to Python3 and TelepathyGLib , and will work on fixing collaboration now.

Saumya-Mishra9129 committed:
Encode str to bytes

write method here used to take unicode in python2
use encoded bytes in place of unicode in python3

Saumya-Mishra9129 commented:
Tested Collaboration, working i.e. game state is shared. However when a player wins and goes to new level, until second player doesn't reach to same level as first, game state is not shared. 
It is shared with both on same levels. I don't know what way the collaboration is desired here .
@quozl @chimosky @pro-panda Please review.

quozl commented:
So to paraphrase, you're not sure how to implement shared game state after the first round?  I think the game state should be the same; perhaps the new level should only occur once both players have won it once.  If a wider discussion is needed, use the issue https://github.com/sugarlabs/cookie-search-activity/issues/18.

Saumya-Mishra9129 commented:
> you're not sure how to implement shared game state after the first round? 

No that is not the case. The way I have implemented collaboration , works only when both players plays on same levels. If one player wins first, collaboration stops and resumes when second player comes on same level as first. however as you said level upgradation should only be possible when both players win. I was saying about perhaps it can be the case that author who have designed the activity , wanted collaboration this way. I am not sure, but it can be possible.

JuiP commented:
I haven't tested collaboration, but is it possible to start a new game when either player wins? This way you can ensure that the game state will always be shared.

chimosky commented on SearchActivity.py:
```diff
@@ -79,7 +66,13 @@ def __init__(self, handle):
+
+        self._collab = CollabWrapper(self)
+        self._collab.connect('message', self._message_cb)
+        self._collab.connect('joined', self._joined_cb)
```
It'll be better to have the joined signal connected from one API to one callback, not having one from two different APIs to same callback.

chimosky commented on SearchActivity.py:
```diff
@@ -187,142 +180,64 @@ def _write_scores_to_clipboard(self, button=None):
         """ ...or join an exisiting share. """
         self._new_tube_common(False)
 
     def _new_tube_common(self, sharer):
```
You can delete this method since you're no longer using it and it'll be better to handle anything that you need to handle when the activity is joined in the `_joined_cb` rather than having a method that calls another method that calls another method to perform an action when that can be handled in the first method.

chimosky replied:
Same thing also happens in `game.py`, you can eliminate the need for those methods by just having `collab.post` calls, one example is [this](https://github.com/sugarlabs/cookie-search-activity/blob/master/game.py#L158) where `send_new_game` is called and that in turn calls `send_event`.

chimosky commented on SearchActivity.py:
```diff
@@ -187,142 +180,64 @@ def _write_scores_to_clipboard(self, button=None):
-                error_handler=self._list_tubes_error_cb)
         self._game.set_sharing(True)
 
     def _list_tubes_reply_cb(self, tubes):
```
This method is also not needed anymore.

chimosky commented on SearchActivity.py:
```diff
@@ -187,142 +180,64 @@ def _write_scores_to_clipboard(self, button=None):
     def _setup_dispatch_table(self):
         ''' Associate tokens with commands. '''
         self._processing_methods = {
             'n': [self._receive_new_game, 'get a new game grid'],
             'p': [self._receive_dot_click, 'get a dot click'],
```
It'll be great to have more informative key names for `n` and `p` like `new_game` and `played`.

chimosky commented on SearchActivity.py:
```diff
@@ -187,142 +180,64 @@ def _write_scores_to_clipboard(self, button=None):
         self._game.remote_button_press(dot, color)
 
-    def send_event(self, entry):
+    def send_event(self, command, payload):
```
You can also get rid of this method and simply have `collab.post` calls where you need them.

Saumya-Mishra9129 replied:
Thanks

chimosky commented:
@Saumya-Mishra9129 said
> No that is not the case. The way I have implemented collaboration , works only when both players plays on same levels. If one player wins first, collaboration stops and resumes when second player comes on same level as first. however as you said level upgradation should only be possible when both players win. I was saying about perhaps it can be the case that author who have designed the activity , wanted collaboration this way. I am not sure, but it can be possible.

Collaboration should work when players are on the same level and when one wins, the other player's game state should be automatically updated and the player should be on the same level as the previous winner.

quozl commented:
> Collaboration should work when players are on the same level and when one wins, the other player's game state should be automatically updated and the player should be on the same level as the previous winner.

I agree.  Thanks for expressing it this way.

Saumya-Mishra9129 committed:
Port to CollabWrapper
